### PR TITLE
allow configuring the ring for a DistributedStore

### PR DESCRIPTION
**Pull request by ccutrer**
either a default ring implementation with custom replicas, or a
fully custom ring

### CODE REVIEW

**Comment by tubbo** [comment]
hey @ccutrer can you check on the test failures?

**Comment by ccutrer** [comment]
Heh, they're failing because they're assuming that https://github.com/redis/redis-rb/pull/739 has been merged in (which changes from using 160 as the default for the parameter, to defaulting it in the method body if nil is passed). I'll re-work this commit to not assume that, since the turnaround time for this project has been much faster in the past than for redis-rb.

**Comment by ccutrer** [comment]
K, updated and should pass tests now. Let me know if you'd prefer to conditionally call HashRing.new with no arguments if replicas isn't passed (allowing it to use the parameter default) instead of assuming what the default should be.

File: lib/redis/distributed_store.rb [hunk 1]
```diff
@@ -6,12 +6,12 @@ class DistributedStore < Distributed
...
-      nodes = addresses.map do |address|
-        ::Redis::Store.new _merge_options(address, options)
-      end
-
       _extend_namespace options
-      @ring = Redis::HashRing.new nodes
+      @ring = options[:ring] || Redis::HashRing.new([], options[:replicas] || Redis::HashRing::POINTS_PER_SERVER)
+
+      addresses.each do |address|
+        @ring.add_node(::Redis::Store.new _merge_options(address, options))
```
Comment (tubbo) [inline lib/redis/distributed_store.rb]: this feels like an unnecessary change to me, can you explain more about what it's doing?

Comment (ccutrer) [reply lib/redis/distributed_store.rb]: if you pass in a fully custom ring implementation, the nodes still need added to that ring. this is how "regular" Redis::Distributed does it as well: https://github.com/redis/redis-rb/blob/master/lib/redis/distributed.rb#L18

Comment (tubbo) [reply lib/redis/distributed_store.rb]: cool, thanks!

Comment (tubbo) [reply lib/redis/distributed_store.rb]: `HashRing` should do that for you: https://github.com/redis/redis-rb/blob/master/lib/redis/hash_ring.rb#L18-L20

File: lib/redis/distributed_store.rb [hunk 2]
```diff
@@ -6,12 +6,12 @@ class DistributedStore < Distributed
     attr_reader :ring
 
     def initialize(addresses, options = { })
-      nodes = addresses.map do |address|
-        ::Redis::Store.new _merge_options(address, options)
-      end
-
       _extend_namespace options
-      @ring = Redis::HashRing.new nodes
+      @ring = options[:ring] || Redis::HashRing.new([], options[:replicas] || Redis::HashRing::POINTS_PER_SERVER)
```
Comment (tubbo) [inline lib/redis/distributed_store.rb]: If `HashRing` took keyword arguments and defaulted to `nodes=[]`, you could instantiate one like so:

```ruby
Redis::HashRing.new(replicas: options[:replicas], )
```

Comment (ccutrer) [reply lib/redis/distributed_store.rb]: Right, but that still doesn't work if I want to pass in my own custom HashRing object (without any nodes populated yet).

say I have a class called RandomlyPutNodesInRingRing

my_ring = RandomlyPutNodesInRingRing.new
redis_store = Redis::Store::Factory.create(nodes_from_my_config, ring: my_ring)

that will pass the ring in, with no nodes in it, and then will create and add the nodes from my config. the Redis::HashRing constructor will never get called

**Comment by ccutrer** [comment]
I added a new PS with some tests for the new functionality

File: lib/redis/distributed_store.rb (hunk 2 above)
Comment (ccutrer) [reply lib/redis/distributed_store.rb]: Doh, I think you understood that. I totally didn't read your comment right. And yes, that would be nice, but it doesn't. So I have to pass an empty array

Comment (tubbo) [reply lib/redis/distributed_store.rb]: no you don't. https://github.com/redis/redis-rb/blob/master/lib/redis/hash_ring.rb#L13

Comment (ccutrer) [reply lib/redis/distributed_store.rb]: I'm not sure what you're asking here. The current state of Redis::HashRing.new is two positional parameters. If I want to pass something to the second parameter (I do!), I have to supply the first parameter, so I'm passing exactly what it defaults to. Yes, it would be nice if it took keyword params instead of positional parameters, so that I could just pass the one arg I care about, but it doesn't. Are you wanting me to try and get a keyword arg pull request accepted to redis-rb?

Comment (tubbo) [reply lib/redis/distributed_store.rb]: i guess that would be kind-of a breaking change.

**Comment by ccutrer** [comment]
Ping?

File: lib/redis/distributed_store.rb (hunk 1 above)
Comment (tubbo) [reply lib/redis/distributed_store.rb]: hah sorry i am an idiot...re-read what `HashRing#initialize` does and you definitely need this!
